src/Bicep.Core/Analyzers/Linter/LinterAnalyzer.cs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Analyzers.Interfaces;
+using Bicep.Core.Semantics;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+
+namespace Bicep.Core.Analyzers.Linter
+{
+    internal class LinterAnalyzer : IBicepAnalyzer
+    {
+        public static string AnalyzerName => "Bicep Internal Linter";
+        public static readonly bool IsCliInvoked;
+
+        private static readonly IEnumerable<Type> RuleTypes;
+        private static readonly ImmutableArray<IBicepAnalyzerRule> RuleSet;
+
+        static LinterAnalyzer()
+        {
+            var entryAssembly = Assembly.GetEntryAssembly();
+            IsCliInvoked = entryAssembly.Location.EndsWith("bicep.dll", true, CultureInfo.CurrentCulture);
+            
+            RuleTypes = Assembly.GetExecutingAssembly()
+                .GetTypes()
+                .Where(t => t.IsSubclassOf(typeof(LinterRule)));
+
+            RuleSet = CreateLinterRules().ToImmutableArray();
+        }
+
+        internal static  IEnumerable<IBicepAnalyzerRule> CreateLinterRules()
+        {
+            foreach(var ruleType in RuleTypes)
+            {
+                if(typeof(IBicepAnalyzerRule).IsAssignableFrom(ruleType))
+                {
+                    yield return (IBicepAnalyzerRule) Activator.CreateInstance(ruleType);
+                }
+            }
+        }
+
+        public IEnumerable<IBicepAnalyzerRule> GetRuleSet() => RuleSet;
+
+        public IEnumerable<IBicepAnalyzerDiagnostic> Analyze(SemanticModel semanticModel)
+            => RuleSet.Where(rule => (IsCliInvoked && rule.EnabledForCLI) || (!IsCliInvoked && rule.EnabledForEditing))
+                .SelectMany(r => r.Analyze(semanticModel));
```

Comment:
Is the decision to use lazy evaluation via `IEnumerable` deliberate? E.g. if someone calls `Analyze(..)` without iterating over the results, the analyzers will not be invoked, and if the results are iterated over multiple times, the analyzers will be invoked multiple times each.

If not, I wonder if it's worth forcing evaluation with a `.ToArray()` to avoid this.

Reply:
enumerating at Analyze now